nemo_skills/inference/prompt/utils.py

```diff
@@ -37,19 +39,92 @@
 
 
 @dataclass
-class PromptConfig:
+class FewShotExamples:
     delimiter: str = MISSING
-    prefix: str = MISSING
     template: str = MISSING
+
+
+@dataclass
+class PromptConfig:
+    few_shot_examples: FewShotExamples = MISSING
+    prompt_template: str = MISSING
+    user: str = MISSING
+    system: str = MISSING
     context_type: str = "empty"
     examples_type: Optional[str] = None
     num_few_shots: int = 5
 
 
+@dataclass
+class Prompt:
+    config: PromptConfig
+    input_dict: Dict[str, Any]
+    example_dicts: List[Dict[str, Any]] = None
+    context_template: Optional[str] = None
+    generated_solution = ""
+
+    def __post_init__(self):
+        """Initialize example_dicts if not provided."""
+        if self.example_dicts is None:
+            self.example_dicts = examples_map.get(self.config.examples_type, [])[: self.config.num_few_shots]
+        if self.context_template is None:
+            self.context_template = context_templates.get(self.config.context_type, "")
+
+    def build_context(self, example_dict: Dict[str, Any]) -> str:
+        """Builds the context string based on the example dictionary."""
+        context = self.context_template.format(**example_dict)
+        return context
+
+    def build_filled_example(self, example_dict: Dict[str, Any]) -> str:
+        """Builds a filled example string based on the example dictionary."""
+        context = self.build_context(example_dict)
+        return self.config.few_shot_examples.template.format(context=context, **example_dict)
+
+    def build_examples(self) -> str:
+        """Builds all examples string concatenated by delimiter."""
+        filled_examples = [self.build_filled_example(example) for example in self.example_dicts]
+        examples = (
+            self.config.few_shot_examples.delimiter.join(filled_examples) + self.config.few_shot_examples.delimiter
+            if filled_examples
+            else ""
+        )
+        context = self.build_context(self.input_dict)
+        user = self.config.user.format(examples=examples, context=context, **self.input_dict)
+        return user
+
+    def build_structured_examples(self) -> List[Dict[str, str]]:
+        """Builds a structured representation of the prompt."""
+        structured_prompt = [{"role": "system", "content": self.config.system}] if self.config.system else []
```

Comment:
Is this how it's supposed to be used? Empty list instead of empty content when system is missing? If empty content works, it will be simpler and more consistent, like this
```suggestion
        structured_prompt = [{"role": "system", "content": self.config.system}]
```

Reply:
No, providing empty system prompt leads to bad generations like this `'generated_solution': '\\boxed{27}'`